### DISCUSSION TITLE
How to disable `Email2FA` action for specific user

### DISCUSSION THREAD

**Discussion by d4f3dQnK** [codeigniter4/shield#525] [Q&A] [answered]
Hello everyone. 

I store in database whether user wants email 2fa authentication. The question is simple, how can I apply this control in class?

Sorry to my poor English. 🤗

**Comment by datamweb** [comment]
Hello,
I didn't understand
Can you explain a little more?

https://lingva.ml/

**Comment by d4f3dQnK** [comment]
For instance, I want to check user's email 2fa option. Like:

```php
        if ($user->email2fa === true && $authenticator->hasAction()) {
            return redirect()->route('auth-action-show')->withCookies();
        }
```

But it don't work like that. I think because sessions already set.

@datamweb Thank you

**Reply by d4f3dQnK** [reply]
*Code example in LoginController

**Reply by datamweb** [reply]
`email2fa` Where did it come from? Have you added a column named `email2fa` to the users table?

Explain what you want to add to the shield?
What made you make these changes?

**Reply by d4f3dQnK** [reply]
I'm so sorry.

I add a column in database. Named: email2fa.

If user set the email2fa setting false in their profile. I want to check on LoginController and pass the Email2FA action. But I can't in LoginController because (i think) auth action sessions are set. 

How can i pass the Email2FA action for each user?

**Reply by datamweb** [reply]
It seems, the possibility of canceling the action through the `LoginController` is currently **not available**.

**Reply by ShapeShifter42** [reply]
> Explain what you want to add to the shield?

I _think_ what he’s looking for is to be able to enable/disable 2FA on a per-user basis. If a user feels that they want the added security of 2FA, they can use it. If they don’t want to be bothered with the extra step, it can be disabled, but only for that particular user. 

There are many sites out there that don’t _require_ 2FA, but they do allow a user to set it up for themselves. It’s a feature I may be interested in - the site I’m designing doesn’t really require it, but it would be nice to offer. On the other hand, many prospective users for my site have very limited computer expertise, and would be hampered or put off if it were forced on them. Some of my users tend to be high maintenance, and I know they would have a lot of trouble if they _had_ to use 2FA.

**Comment by kenjis** [comment, marked as answer]
Extend the `Email2FA` class so that `createIdentity()` does nothing and returns if `$user->email2fa` is false.

**Reply by d4f3dQnK** [reply]
Thank you so much!

**Reply by AndrewDFrazier-SwofMedia** [reply]
The createIdentity() function is final and throws a fatal error if you try to override it: "Fatal error: Cannot override final method CodeIgniter\Shield\Authentication\Actions\Email2FA::createIdentity()".  How would one go about getting around this?

**Reply by kenjis** [reply]
Upgrade your Shield.

**Reply by santosh-techmarcos** [reply]
where can i extend that Email2FA class

**Comment by kenjis** [comment]
It is better to extend `Session::startUpAction()` so that it returns false.

**Reply by santosh-techmarcos** [reply]
How? and which file i have to extent that class

**Reply by kenjis** [reply]
This.
https://github.com/codeigniter4/shield/blob/a38b3ccc5d70862ffc27b470a76cb4156f4d1a70/src/Authentication/Authenticators/Session.php#L197

**Reply by santosh-techmarcos** [reply]
what i do if i want to skip Email2FA for admin's there are any way to achived that functionality.

**Reply by santosh-techmarcos** [reply]
can you please give me a sample whre i extend StartUpAction class.

**Reply by kenjis** [reply]
Add code to do nothing when an admin user logs in in your `Session::startUpAction()` method.
Or extend the `Email2FA` class so that `createIdentity()` does nothing and returns.

Do what you can do.
